isoutils/isoutils.go

```diff
@@ -0,0 +1,539 @@
+	kernelType := strings.Split(kernelTypeVersion, ".")[0] //kernelType examples: native,kvm,lts2018,hyperv
+	kernelVersion := strings.SplitAfter(kernelTypeVersion, kernelType+".")[1]
+
+	/* Copy files to initrd, and add to mods so they're added to the init template */
+	for _, i := range modules {
+		rootfsModPath := tmpPaths[clr_rootfs] + "/usr/lib/modules/" + kernelVersion + "." + kernelType + i
+
+		/* copy kernel module to initramfs */
+		initrdModPath := filepath.Dir(tmpPaths[clr_initrd] + "/usr/lib/modules/" + kernelVersion + "." + kernelType + i)
+
+		if _, err := os.Stat(initrdModPath); os.IsNotExist(err) {
+			os.MkdirAll(initrdModPath, os.ModePerm)
+		}
+
+		err = utils.CopyFile(rootfsModPath, initrdModPath+"/"+filepath.Base(i))
+		if err != nil {
+			prg.Failure()
+			return err
+		}
+		mods.Modules = append(mods.Modules, "/usr/lib/modules/"+kernelVersion+"."+kernelType+i)
+	}
+
+	tmpl, err := ioutil.ReadFile(templatePath + "/initrd_init_template")
+	if err != nil {
+		prg.Failure()
+		log.Error("Failed to open: initrd template at %s\n", templatePath+"initrd_init_template")
+		return err
+	}
+
+	t := template.New("Modules template")
+	t, err = t.Parse(string(tmpl))
+	if err != nil {
+		prg.Failure()
+		log.Error("Failed to parse init's template")
+		return err
+	}
+
+	f, err := os.Create(tmpPaths[clr_initrd] + "/init")
+	if err != nil {
+		prg.Failure()
+		log.Error("Failed to create init file for initrd!")
+		return err
+	}
+	defer f.Close()
+
+	err = t.Execute(f, mods)
+	if err != nil {
+		prg.Failure()
+		log.Error("Failed to execute template filling")
+		return err
+	}
+
+	/* Set correct owner and permissions on initrd's init */
+	if err = os.Chown(tmpPaths[clr_initrd]+"/init", 0, 0); err != nil {
+		prg.Failure()
+		return err
+	}
+	if err = os.Chmod(tmpPaths[clr_initrd]+"/init", 0700); err != nil {
+		prg.Failure()
+		return err
+	}
+
+	prg.Success()
+	return err
+}
+
+/* Build initrd image, and copy to the correct location */
+func buildInitrdImage() error {
+	msg := "Building initrd image"
+	prg := progress.NewLoop(msg)
+	log.Info(msg)
+
+	// Determine executible's path so we can revert to it when this function ends
+	ex, err := os.Executable()
+	if err != nil {
+		prg.Failure()
+		return err
+	}
+	exPath := filepath.Dir(ex)
+
+	/* find all files in the initrd path, create the initrd */
+	/* The find command must return filenames without a path (eg, must be run in the current dir) */
+	os.Chdir(tmpPaths[clr_initrd])
+
+	args := "sudo find .| cpio -o -H newc | gzip >" + tmpPaths[clr_cdroot] + "/images/initrd.gz"
+	_, err = exec.Command("bash", "-c", args).CombinedOutput()
+	if err != nil {
+		prg.Failure()
+		return err
+	}
+	os.Chdir(exPath)
+	prg.Success()
+	return err
+}
+
+func mkEfiBoot() error {
+	msg := "Building efiboot image"
+	prg := progress.NewLoop(msg)
+	log.Info(msg)
+
+	cmds := [][]string{
+		[]string{"fallocate", "-l", "800M", tmpPaths[clr_cdroot] + "/EFI/efiboot.img"},
```

Comment:
This really doesn't need to be 800M, I think I got this size from the previous ISO script.

Currently the size of the *FILES* installed to this image is 32MB, I'd suggest probably 100MB?
Or maybe there's some way to resize this FS to the content size after creation?

The current ISO for a live image is 2.7GB. When xz'd the size is 1.5GB. Most of this compression is compressing this huge, empty EFI boot partition.